File: cpp/Array/L_insertDeleteRandom.cpp
#include <bits/stdc++.h>
using namespace std;

class RandomizedSet
{
private:
    vector<int> a;
    unordered_map<int, int> m;

public:
    RandomizedSet()
    {
    }

    bool insert(int val)
    {
        if (m.find(val) != m.end())
            return false;
        else
        {
            a.push_back(val);
            m[val] = a.size() - 1;
            return true;
        }
    }

    bool remove(int val)
    {
        if (m.find(val) == m.end())
            return false;
        else
        {
            int last = a.back();
            a[m[val]] = a.back();
            a.pop_back();
            m[last] = m[val];
            m.erase(val);
            return true;
        }
    }

    int getRandom()
    {
        return a[rand() % a.size()];
    }
};

class RandomizedSet
{
    unordered_map<int, int> S;
    vector<int> idx;
    int sz = 0;
    mt19937 rng; // Mersenne Twister random number engine
public:
    RandomizedSet()
    {
        rng = mt19937(time(NULL));
    }

    bool insert(int val)
    {
        if (S.count(val))
            return 0;

        S[val] = sz++;
        idx.push_back(val);
        return 1;
    }

    bool remove(int val)
    {
        if (!S.count(val))
            return 0;

        int r = idx.back();
        int pos = S[val];

        S[r] = pos;
        idx[pos] = r;

        S.erase(val);
        idx.pop_back();
        sz--;

        return 1;
    }
    int getRandom()
    {
        return idx[rng() % sz];
    }
};

auto init = []()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 'c';
}();

File: Python/Arrays/L_insertDeleteRandom.py
import random
from operator import setitem


class RandomizedSet:
    
    def __init__(self):
        self.dataMap = {}
        self.data = []

    def insert(self, val: int) -> bool:
        if val in self.dataMap:
            return False
        self.dataMap[val] = len(self.data)
        self.data.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.dataMap:
            return False
        lastElement = self.data[-1]
        idxToRemove = self.dataMap[val]
        
        self.dataMap[lastElement] = idxToRemove
        self.data[idxToRemove] = lastElement
        self.data[-1] = val
        self.data.pop()
        self.dataMap.pop(val)
        return True
    
    def getRandom(self) -> int:
        return random.choice(self.data)
    
    
class RandomizedSet:
    def __init__(self):
        self.a,self.d=[],{}


    def insert(self, v: int) -> bool:
        return (v in self.d)^1 and not (setitem(self.d,v,len(self.a)) or self.a.append(v))
        

    def remove(self, v: int) -> bool:
        return v in self.d and not (lambda e,i:i<len(self.a) and (setitem(self.a,i,e) or setitem(self.d,e,i)))(self.a.pop(),self.d.pop(v))
        

    def getRandom(self) -> int:
        return choice(self.a)

